day17/go/razziel89/README.adoc

```diff
@@ -0,0 +1,63 @@
+== Day 17: Trick Shot ==
+
+This is my implementation for both rounds of the trick shot puzzle.
+
+This one was actually really interesting in Golang.
+I've never had to construct an exhaustive list of velocities or an exhaustive
+list of trajectories.
+Instead, I chained one goroutine that generates velocities and another one that
+analyses trajectories based on them.
+That second one only provides solutions that are acceptable, which I count and
+whose maximum achievable height I track.
+
+The problem with this problem (haha) is, that we never know whether another,
+higher starting velocity in y direction won't actually still allow us to reach
+the target area.
+That is, technically, a velocity of x=5, y=1million could still hit the target
+area.
+At least I could not come up with a quick way to assess when to stop.
```

Comment:
Thank you for sharing your solutions including reasoning, I enjoy reading them!

My reasoning for when to stop was as follows: the way up and the way down are symmetric. If the projectile hits a y-coordinate y_0 on the way up, with a velocity of y_vel_0, then it will also hit that y-coordinate on the way down, and will have a velocity of -y_vel_0. 

So, for the case where the target area is below the initial position (so, y_min < y_max < 0),  the maximal viable y-velocity is -y_min. Because of symmetry, the projectile will pass y_0=0 on the way down, with a velocity of y_min -- so it will get into the target area in the next step (assuming it hits the x-coordinates). If you start with a higher velocity, you will overshoot.

If the target area is above the start position, it's easier -- maximal velocity is y_max, otherwise you will overshoot in the first step, and because of symmetry also won't hit on the way down.

With this, you can also solve the first part of the problem without code. If you shoot the projectile with y-velocity k, it will peak at k * (k+1) / 2, or the sum of numbers from 1 to k.

Reply:
> Thank you for sharing your solutions including reasoning, I enjoy reading them!

Thanks for the kind words! I am very happy to see that someone had a look at them. I also like reading other people's solutions a lot.

> So, for the case where the target area is below the initial position (so, y_min < y_max < 0), the maximal viable y-velocity is -y_min. Because of symmetry, the projectile will pass y_0=0 on the way down, with a velocity of y_min -- so it will get into the target area in the next step (assuming it hits the x-coordinates). If you start with a higher velocity, you will overshoot.

That's very true, thanks for the explanation! I now see that the target areas are always fully below y==0. That's something I didn't see or at least I thought that all the constraints would be given in the task. It's a good idea to have a look at the actual data and try to derive constraints that simplify the solution :+1:. Just writing down my current thoughts: Actually, due to the symmetry of the trajectory in y direction, if the area overlaps with y==0, there will be no maximum possible height since a projectile starting at y==0 will always come back to that height.

> If the target area is above the start position, it's easier -- maximal velocity is y_max, otherwise you will overshoot in the first step, and because of symmetry also won't hit on the way down.

True, the max viable y velocity then is `max(y_area_max, -y_area_min)`.

Thanks for sharing your thoughts on this, I really appreciate it!